rasa_nlu/extractors/nested_entity_extractor.py

```diff
@@ -0,0 +1,264 @@
+from typing import Text
+
+from rasa_nlu import utils
+from rasa_nlu.extractors import EntityExtractor
+from rasa_nlu.training_data import Message
+from rasa_nlu.training_data import TrainingData
+from rasa_nlu.utils import write_json_to_file
+
+from word2number import w2n
+
+try:
+    # Python 2
+    from __builtin__ import str as builtin_str
+except ImportError:
+    # Python 3
+    from builtins import str as builtin_str
+
+NESTED_ENTITIES_FILE_NAME = "nested_entities.json"
+
+
+class NestedEntityExtractor(EntityExtractor):
+    name = "nested_entity_extractor"
+    provides = ["entities"]
+
+    def __init__(self, component_config=None, nested_entities=None):
+        # type: (Optional[Dict[Text, Text]]) -> None
+        super(NestedEntityExtractor, self).__init__(component_config)
+
+        self.nested_entities = nested_entities if nested_entities else {
+            'lookup_tables': [],
+            'composite_entities': []
+        }
+
+    def train(self, training_data, cfg, **kwargs):
+        # type: (TrainingData) -> None
+        self.add_lookup_tables(training_data.lookup_tables)
+        ce = training_data.composite_entities
+        self.nested_entities['composite_entities'] = ce
+
+    def process(self, message, **kwargs):
+        # type: (Message, **Any) -> None
+        entities = message.get("entities", [])[:]
+        self.split_nested_entities(entities)
+        message.set("entities", entities, add_to_output=True)
+
+    def persist(self, model_dir):
+        # type: (Text) -> Optional[Dict[Text, Any]]
+        if self.nested_entities:
+            nested_entities_file = os.path.join(model_dir,
+                                                NESTED_ENTITIES_FILE_NAME)
+            write_json_to_file(nested_entities_file, self.nested_entities,
+                               separators=(',', ': '))
+
+        return {"nested_entities_file": NESTED_ENTITIES_FILE_NAME}
+
+    @classmethod
+    def load(cls,
+             model_dir=None,  # type: Optional[Text]
+             model_metadata=None,  # type: Optional[Metadata]
+             cached_component=None,  # type: Optional[NestedEntitiesMapper]
+             **kwargs  # type: **Any
+             ):
+            # type: (...) -> NestedEntitiesMapper
+
+        meta = model_metadata.for_component(cls.name)
+        file_name = meta.get("nested_entities_file", NESTED_ENTITIES_FILE_NAME)
+        nested_entities_file = os.path.join(model_dir, file_name)
+
+        if os.path.isfile(nested_entities_file):
+            nested_entities = utils.read_json_file(nested_entities_file)
+        else:
+            nested_entities = {
+                'lookup_tables': [],
+                'composite_entities': []
+            }
+            warnings.warn("Failed to load nested entities file from '{}'"
+                          "".format(nested_entities_file))
+
+        return cls(meta, nested_entities)
+
+    def get_relevance(self,
+                      broad_value,
```

Comment:
Why `broad_value` and not `parent`?

Reply:
`parent` will work but this functions get called all round the place hence I thought that passing round the entire word that I am breaking all round will make more sense hence why I called it `broad_value`